include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,844 @@
+  * @param status            Failure code if unsuccessful
+  *
+  * @retval 0 on success
+  * @retval <0 on failure
+  */
+int odp_comp_session_create(odp_comp_session_param_t *param,
+			    odp_comp_session_t *session,
+			    odp_comp_ses_create_err_t *status);
+
+ /**
+  * Comp session destroy
+  *
+  * Destroy an unused session. Result is undefined if session is being used
+  * (i.e. asynchronous operation is in progress).
+  *
+  * @param session           Session handle
+  *
+  * @retval 0 on success
+  * @retval <0 on failure
+  */
+int odp_comp_session_destroy(odp_comp_session_t session);
+
+/**
+ * Comp set dictionary
+ *
+ * Should be called when there is no operation in progress i.e.
+ * before initiating processing of first chunk of data and
+ * after processing of last chunk of data is complete.
+ *
+ * @param session           Session handle
+ * @param dict[in,out]      Pointer to dictionary.
+ *                          implementation should update length of dictionary
+ *                          used at output.
+ * @retval 0 on success
+ * @retval <0 on failure
+ *
+ * @note:
+ * Application should call odp_comp_alg_capability() to query 'support_dict'
+ * before making call to this API.
+ */
+int odp_comp_set_dict(odp_comp_session_t session,
+		      odp_comp_dict_t *dict);
+
+/**
+ * Comp packet operation synchronous
+ *
+ * Performs the SYNC compression/decompression operations on the packets.
+ *
+ * If session is created in ODP_COMP_SYNC mode, this call wait for operation
+ * to complete.
+ *
+ * If session is created in ODP_COMP_ASYNC mode, behavior is undefined.
+ *
+ * for compression + hash, call returns with hash appended to the end of
+ * last processed chunk of data.
+ *
+ * User should compute processed data len = total output len - digest_len, where
+ * digest_len queried through odp_comp_hash_alg_capability().
+ *
+ * Caller should initialize pkt_out with valid output
+ * packet handles. All arrays should be of num_pkt size.
+ *
+ * In case of partially processed array i.e. when
+ * number of packets returned < num_pkt, application should call
+ * odp_comp_result() on 1st failed packet (i.e. n+1th packet where
+ * n=number of packets returned) to ensure if error code is
+ * ODP_COMP_ERR_OUT_OF_SPACE.
+ *
+ * In case of ODP_COMP_ERR_OUT_OF_SPACE, application should keep on calling
+ * odp_comp_op() API with more output buffer until call returns
+ * with num_pkt or with result as ODP_COMP_ERR_NONE or any other
+ * failure code.
+ *
+ * @param         pkt_in   Packets to be processed
+ * @param[in,out] pkt_out  Packet handle array specifying resulting packets
+ * @param         param    Operation parameters array
+ * @param         num_pkt  Number of packets to be processed
+ *
+ * @return Number of input packets consumed (0 ... num_pkt)
+ * @retval <0 on failure
+ */
+int odp_comp_compress(const odp_packet_t pkt_in[],
+		      odp_packet_t pkt_out[],
+		      odp_comp_packet_op_param_t param[],
```

Comment:
Is there any reason why `param` is not `const`?

Reply:
we will be updating length of data written into output data range, thus , its not constant. 
BTW, it reminds me a question in crypto pkt_out [] is constant array, don't you update it if user input ODP_PACKET_INVALID and crypto do auto-allocation

Reply:
`pkt_out` is updated during sync call and it serves only as a reference during async call.
I'd suggest to move resulting length to compression result, rather than updating compression param. See, parameters might be unavailable when resulting packets are returned from async call.

Reply:
Even for sync? if yes, then I have an open question on new design that do user need to call odp_crypto_result() to get per packet result after odp_crypto_op().. is over? if yes, then it means one sync call translates to two calls? to get final o/p?